mediapipe/tasks/ios/genai/core/sources/LlmTaskRunner.swift

```diff
@@ -53,25 +52,99 @@ public final class LlmTaskRunner {
+    /// being invoked. `CallbackInfo` takes care of freeing the memory of `inputText` when it is
+    /// deallocated.
+    let callbackInfo = CallbackInfo(
+      inputText: strdup(inputText), progress: progress, completion: completion)
+    let callbackContext = UnsafeMutableRawPointer(Unmanaged.passRetained(callbackInfo).toOpaque())
+
+    LlmInferenceEngine_Session_PredictAsync(cLlmSession, callbackContext, callbackInfo.inputText) {
+      context, responseContext in
+      guard let cContext = context else {
+        return
+      }
+
+      /// `takeRetainedValue()` decrements the reference count incremented by `passRetained()`. Only
+      /// take a retained value if the LLM has finished generating responses to prevent the context
+      /// from being deallocated in between response generation.
+      let cCallbackInfo =
+        responseContext.done
+        ? Unmanaged<CallbackInfo>.fromOpaque(cContext).takeRetainedValue()
+        : Unmanaged<CallbackInfo>.fromOpaque(cContext).takeUnretainedValue()
+
+      if let responseStrings = LlmTaskRunner.responseStrings(from: responseContext) {
+        cCallbackInfo.progress(responseStrings, nil)
+      } else {
+        cCallbackInfo.progress(nil, GenAiInferenceError.invalidResponse)
+      }
+
+      /// Call completion callback if LLM has generated its last response.
+      if responseContext.done {
+        cCallbackInfo.completion()
+      }
+    }
+  }
+
+  deinit {
+    LlmInferenceEngine_Session_Delete(cLlmSession)
+  }
+}
+
+extension LlmTaskRunner {
+  /// A wrapper class for whose object will be used as the C++ callback context.
+  /// The progress and completion callbacks cannot be invoked without a context.
+  class CallbackInfo {
+    typealias ProgressCallback = (_ partialResult: [String]?, _ error: Error?) -> Void
+    typealias CompletionCallback = () -> Void
+
+    let inputText: UnsafeMutablePointer<CChar>?
+    let progress: ProgressCallback
+    let completion: CompletionCallback
+
+    init(
+      inputText: UnsafeMutablePointer<CChar>?, progress: @escaping (ProgressCallback),
+      completion: @escaping (CompletionCallback)
+    ) {
+      self.inputText = inputText
+      self.progress = progress
+      self.completion = completion
+    }
+
+    deinit {
+      free(inputText)
+    }
+  }
+}
+
+extension LlmTaskRunner {
+  private class func responseStrings(from responseContext: LlmResponseContext) -> [String]? {
     guard let cResponseArray = responseContext.response_array else {
-      throw GenAiInferenceError.invalidResponseError
+      return nil
     }
 
     var responseStrings: [String] = []
-
     for responseIndex in 0..<Int(responseContext.response_count) {
+      /// Throw an error if the response string is `NULL`.
       guard let cResponseString = cResponseArray[responseIndex] else {
-        throw GenAiInferenceError.invalidResponseError
+        return nil
```

Comment:
Same here. Also here it doesn't correspond to the comment set above, so should I remove this comment altogether?